File: android/src/main/java/com/waassdkreactnative/MPCKeyService.java
package com.waassdkreactnative;

import static com.waassdkinternal.v1.V1.newMPCKeyService;
import static com.waassdkreactnative.Utils.convertJsonToArray;
import static com.waassdkreactnative.Utils.convertMapToJson;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.common.StandardCharsets;
import com.facebook.react.module.annotations.ReactModule;
import com.waassdkinternal.v1.Device;
import com.waassdkinternal.v1.Signature;
import com.waassdkinternal.v1.SignedTransaction;

import org.json.JSONArray;
import org.json.JSONObject;

@ReactModule(name = MPCKeyService.NAME)
public class MPCKeyService extends ReactContextBaseJavaModule {
  public static final String NAME = "MPCKeyService";
  
  // The URL of the MPCKeyService.
  public static final String  mpcKeyServiceUrl = "https://api.developer.coinbase.com/waas/mpc_keys";

  // The error code for MPCKeyService-related errors.
  private String mpcKeyServiceErr = "E_MPC_KEY_SERVICE";

  // The error message for calls made without initializing SDK.
  private String uninitializedErr = "MPCKeyService must be initialized";

  // The handle to the Go MPCKeyService client.
  com.waassdkinternal.v1.MPCKeyService keyClient;

   MPCKeyService(ReactApplicationContext reactContext) {
    super(reactContext);

  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }


/********MPCKeyService SERVICE API'S********************* */

/**
  Initializes the MPCKeyService  with the given parameters.
  Resolves with the string "success" on success; rejects with an error otherwise.
  */
@ReactMethod
public void initialize(String apiKeyName,String privateKey,Promise promise) {
  try{
    keyClient = newMPCKeyService(mpcKeyServiceUrl,apiKeyName,privateKey);
    promise.resolve("success initialize MPC key service");
  } catch(Exception e) {
    promise.reject("initialize MPC key service failed : ", e);
  }
}

/**
  Registers the current Device. Resolves with the Device object on success; rejects with an error otherwise.
  */
@ReactMethod
public void registerDevice(Promise promise) {
  try {

    if(keyClient==null)
    {
      promise.reject(mpcKeyServiceErr,uninitializedErr);
    }

    Device device = keyClient.registerDevice();

    WritableMap map = Arguments.createMap();
    map.putString("Name", device.getName());

    promise.resolve(map);

  } catch(Exception e) {
    promise.reject("registerDevice failed : ", e);
  }
}

/**
  Polls for pending DeviceGroup (i.e. CreateDeviceGroupOperation), and returns the first set that materializes.
  Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
  stopPollingForPendingDeviceGroup or computeMPCOperation) before another call is made to this function.
  Resolves with a list of the pending CreateDeviceGroupOperations on success; rejects with an error otherwise.
  */
  @ReactMethod
  public void pollForPendingDeviceGroup(String deviceGroup,int pollInterval,Promise promise) {
    try{
      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }

      byte[] pendingDeviceGroupData = keyClient.pollPendingDeviceGroup(deviceGroup,pollInterval);

      // Converting the bytes to String.
      String pendingDeviceGroupDataBytesToStrings = new String(pendingDeviceGroupData, StandardCharsets.UTF_8);
      JSONArray jsonArray=new JSONArray(new String(pendingDeviceGroupDataBytesToStrings));

      WritableArray array = convertJsonToArray(jsonArray);
      promise.resolve(array);

    } catch(Exception e) {
      promise.reject("pollForPendingDeviceGroup failed : ", e);
    }
  }

  /**
  Stops polling for pending DeviceGroup. This function should be called, e.g., before your app exits,
  screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending DeviceGroup.
  Resolves with string "stopped polling for pending DeviceGroup" if polling is stopped successfully;
  resolves with the empty string otherwise.
  */
  @ReactMethod
  public void stopPollingPendingDeviceGroup(Promise promise) {
    try{
      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }

      String result = keyClient.stopPollingPendingDeviceGroup();

      promise.resolve(result);

    } catch(Exception e) {
      promise.reject("stopPollingPendingDeviceGroup failed : ", e);
    }
  }

  /**
  Initiates an operation to create a Signature resource from the given transaction.
  Resolves with the string "success" on successful initiation; rejects with an error otherwise.
  */
  @ReactMethod
  public void createSignatureFromTx(String parent, ReadableMap transaction, Promise promise) {
    try{

      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }

      JSONObject serializedTx = convertMapToJson(transaction);
      keyClient.createTxSignature(parent,serializedTx.toString().getBytes(StandardCharsets.UTF_8));

      promise.resolve("success");

    } catch(Exception e) {
      promise.reject("createSignatureFromTx failed : ", e);
    }
  }

  /**
  Polls for pending Signatures (i.e. CreateSignatureOperations), and returns the first set that materializes.
  Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
  stopPollingForPendingSignatures or processPendingSignature before another call is made to this function.
  Resolves with a list of the pending Signatures on success; rejects with an error otherwise.
  */
  @ReactMethod
  public void pollForPendingSignatures(String deviceGroup,int pollInterval,Promise promise) {
    try{
      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }

      byte[] pendingSeedsData = keyClient.pollPendingSignatures(deviceGroup,pollInterval);

      // Converting the bytes to String.
      String pendingSeedsDataBytesToStrings = new String(pendingSeedsData,StandardCharsets.UTF_8);
      JSONArray jsonArray=new JSONArray(new String(pendingSeedsDataBytesToStrings));

      WritableArray array = convertJsonToArray(jsonArray);

      promise.resolve(array);

    } catch(Exception e) {
      promise.reject("pollForPendingSignatures failed : ", e);
    }
  }

  /**
  Stops polling for pending Signatures This function should be called, e.g., before your app exits,
  screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending Signatures.
  Resolves with string "stopped polling for pending Signatures" if polling is stopped successfully;
  resolves with the empty string otherwise.
  */
  @ReactMethod
  public void stopPollingForPendingSignatures(Promise promise) {
    try{
      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }

      String result = keyClient.stopPollingPendingSignatures();

      promise.resolve(result);

    } catch(Exception e) {
      promise.reject("stopPollingPendingSignatures failed : ", e);
    }
  }

/**
  Waits for a pending Signature with the given operation name. Resolves with the Signature object on success;
  rejects with an error otherwise.
  */
  @ReactMethod
  public void waitPendingSignature(String operation, Promise promise) {
    try{

      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }
      Signature signature = keyClient.waitPendingSignature(operation);

      WritableMap map = Arguments.createMap();
      map.putString("Name", signature.getName());
      map.putString("Payload", signature.getPayload());
      map.putString("SignedPayload", signature.getSignedPayload());

      promise.resolve(map);

    } catch(Exception e) {
      promise.reject("waitPendingSignature failed : ", e);
    }
  }

  /**
  Gets the signed transaction using the given inputs.
  Resolves with the SignedTransaction on success; rejects with an error otherwise.
  */
  @ReactMethod
  public void getSignedTransaction(ReadableMap transaction,ReadableMap signature,Promise promise) {
    try{

      if(keyClient==null)
      {
        promise.reject(mpcKeyServiceErr,uninitializedErr);
      }

      JSONObject serializedTx = convertMapToJson(transaction);
      Signature goSignature = new Signature();
      goSignature.setName(signature.getString("Name"));
      goSignature.setPayload(signature.getString("Payload"));
      goSignature.setSignedPayload(signature.getString("SignedPayload"));

      SignedTransaction signedTransaction = keyClient.getSignedTransaction(serializedTx.toString().getBytes(StandardCharsets.UTF_8),goSignature);

      WritableMap map = Arguments.createMap();
      map.putMap("Transaction", transaction);
      map.putMap("Signature", signature);
      map.putString("RawTransaction", signedTransaction.getRawTransaction());
      map.putString("TransactionHash", signedTransaction.getTransactionHash());

      promise.resolve(map);

    } catch(Exception e) {
      promise.reject("getSignedTransaction failed : ", e);
    }
  }

/********END MPCKeyService  API'S********************* */

}

File: ios/swift/WaasSdk/MPCKeyService.swift
import Foundation
import WaasSdkGo
import Combine

public class MPCKeyService: NSObject {
    // The URL of the MPCKeyService.
    let mpcKeyServiceUrl = "https://api.developer.coinbase.com/waas/mpc_keys"

    // The error code for MPCKeyService-related errors.
    let mpcKeyServiceErr = "E_MPC_KEY_SERVICE"

    // The error message for calls made without initializing SDK.
    let uninitializedErr = "MPCKeyService must be initialized"

    // The handle to the Go MPCKeyService client.
    var keyClient: V1MPCKeyServiceProtocol?

    /**
     Initializes the MPCKeyService  with the given parameters.
     Resolves with the string "success" on success; rejects with an error otherwise.
     */
    public init(_ apiKeyName: String, privateKey: String) throws {
        var error: NSError?
        keyClient = V1NewMPCKeyService(
            mpcKeyServiceUrl,
            apiKeyName,
            privateKey,
            &error)

        if error != nil {
            throw WaasError.mpcKeyServiceFailedToInitialize
        }
    }

    private func wrapError(err: Error) -> WaasError {
        return WaasError.mpcKeyServiceUnspecifiedError(err as NSError)
    }

    /**
     Registers the current Device. Resolves with the Device object on success; rejects with an error otherwise.
     */
    public func registerDevice() -> Future<Device, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let device = try self.keyClient?.registerDevice()
                    promise(Result.success(Device(Name: device!.name)))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Polls for pending DeviceGroup (i.e. CreateDeviceGroupOperation), and returns the first set that materializes.
     Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
     stopPollingForPendingDeviceGroup or computeMPCOperation) before another call is made to this function.
     Resolves with a list of the pending CreateDeviceGroupOperations on success; rejects with an error otherwise.
     */
    public func pollForPendingDeviceGroup(_ deviceGroup: String, pollInterval: NSNumber) -> Future<[CreateDeviceGroupOperation], WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let pendingDeviceGroupData = try self.keyClient?.pollPendingDeviceGroup(
                        deviceGroup,
                        pollInterval: pollInterval.int64Value)
                    let people = try JSONDecoder().decode([CreateDeviceGroupOperation].self, from: pendingDeviceGroupData!)
                    promise(Result.success(people))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Stops polling for pending DeviceGroup. This function should be called, e.g., before your app exits,
     screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending DeviceGroup.
     Resolves with string "stopped polling for pending DeviceGroup" if polling is stopped successfully;
     resolves with the empty string otherwise.
     */
    public func stopPollingForPendingDeviceGroup() -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.stopPollingPendingDeviceBackups(goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
         }
    }

    /**
     Initiates an operation to create a Signature resource from the given transaction.
     Resolves with the resource name of the WaaS operation creating the Signature on successful initiation; rejects with an error otherwise.
     */
    public func createSignatureFromTx(_ parent: String, transaction: NSDictionary) -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let serializedTx = try JSONSerialization.data(withJSONObject: transaction)
                    self.keyClient?.createTxSignature(parent, tx: serializedTx, receiver: goReturnsString(promise: promise, wrapAsError: self.wrapError))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Polls for pending Signatures (i.e. CreateSignatureOperations), and returns the first set that materializes.
     Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
     stopPollingForPendingSignatures or computeMPCOperaton before another call is made to this
     function. Resolves with a list of the pending Signatures on success; rejects with an error otherwise.
     */
    public func pollForPendingSignatures(_ deviceGroup: String, pollInterval: NSNumber) -> Future<[PendingSignature], WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let pendingSignaturesData = try self.keyClient?.pollPendingSignatures(
                        deviceGroup,
                        pollInterval: pollInterval.int64Value)
                    let res = try JSONDecoder().decode([PendingSignature].self, from: pendingSignaturesData!)
                    promise(Result.success(res))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Stops polling for pending Signatures This function should be called, e.g., before your app exits,
     screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending Signatures.
     Resolves with string "stopped polling for pending Signatures" if polling is stopped successfully;
     resolves with the empty string otherwise.
     */
    public func stopPollingForPendingSignatures() -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.stopPollingPendingSignatures(goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }

    /**
     Waits for a pending Signature with the given operation name. Resolves with the Signature object on success;
     rejects with an error otherwise.
     */
    public func waitPendingSignature(_ operation: String) -> Future<V1Signature, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                var signature: V1Signature?
                do {
                    signature = try self.keyClient?.waitPendingSignature(operation)
                    promise(Result.success(signature!))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }
    /**
     Gets the signed transaction using the given inputs.
     Resolves with the SignedTransaction on success; rejects with an error otherwise.
     */
    public func getSignedTransaction(_ transaction: NSDictionary, signature: NSDictionary) -> Future<V1SignedTransaction, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let serializedTx = try JSONSerialization.data(withJSONObject: transaction)
                    let goSignature = V1Signature()
                    // swiftlint:disable force_cast
                    goSignature.name = signature["Name"] as! String
                    goSignature.payload = signature["Payload"] as! String
                    goSignature.signedPayload = signature["SignedPayload"] as! String
                    // swiftlint:enable force_cast

                    let signedTransaction = try self.keyClient?.getSignedTransaction(serializedTx, signature: goSignature)
                    promise(Result.success(signedTransaction!))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Gets a DeviceGroup with the given name. Resolves with the DeviceGroup object on success; rejects with an error otherwise.
     */
    public func getDeviceGroup(_ name: String) -> Future<V1DeviceGroup, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let deviceGroupRes = try self.keyClient?.getDeviceGroup(name)
                    promise(Result.success(deviceGroupRes!))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Initiates an operation to prepare device archive for MPCKey export. Resolves with the resource name of the WaaS operation creating the Device Archive on successful initiation; rejects with
     an error otherwise.
     */
    public func prepareDeviceArchive(_ deviceGroup: String, device: String) -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.prepareDeviceArchive(
                    deviceGroup, device: device, receiver: goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }

    /**
     Polls for pending DeviceArchives (i.e. DeviceArchiveOperations), and returns the first set that materializes.
     Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
     stopPollingForDeviceArchives or computePrepareDeviceArchiveMPCOperation) before another call is made to this function.
     Resolves with a list of the pending DeviceArchives on success; rejects with an error otherwise.
     */
    public func pollForPendingDeviceArchives(_ deviceGroup: String, pollInterval: NSNumber) -> Future<[PendingDeviceArchive], WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let pendingDeviceArchiveData = try self.keyClient?.pollPendingDeviceArchives(
                        deviceGroup,
                        pollInterval: pollInterval.int64Value)
                    let res = try JSONDecoder().decode([PendingDeviceArchive].self, from: pendingDeviceArchiveData!)
                    promise(Result.success(res))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Stops polling for pending DeviceArchive operations. This function should be called, e.g., before your app exits,
     screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending DeviceArchiveOperation.
     Resolves with string "stopped polling for pending Device Archives" if polling is stopped successfully; resolves with the empty string otherwise.
     */
    public func stopPollingForPendingDeviceArchives() -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.stopPollingPendingDeviceArchives(goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }

    /**
     Initiates an operation to prepare device backup to add a new Device to the DeviceGroup. Resolves with the resource name of the WaaS operation creating the Device Backup on
     successful initiation; rejects with an error otherwise.
     */
    public func prepareDeviceBackup(_ deviceGroup: String, device: String) -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.prepareDeviceBackup(
                    deviceGroup, device: device, receiver: goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }

    /**
     Polls for pending DeviceBackups (i.e. DeviceBackupOperations), and returns the first set that materializes.
     Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
     stopPollingForDeviceBackups or computePrepareDeviceBackupMPCOperation) before another call is made to this function.
     Resolves with a list of the pending DeviceBackups on success; rejects with an error otherwise.
     */
    public func pollForPendingDeviceBackups(_ deviceGroup: String, pollInterval: NSNumber) -> Future<[PendingDeviceBackup], WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let pendingDeviceBackupData = try self.keyClient?.pollPendingDeviceBackups(
                        deviceGroup,
                        pollInterval: pollInterval.int64Value)
                    // swiftlint:disable force_cast
                    let res = try JSONDecoder().decode([PendingDeviceBackup].self, from: pendingDeviceBackupData!)
                    // swiftlint:enable force_cast
                    promise(Result.success(res))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Stops polling for pending DeviceBackup operations. This function should be called, e.g., before your app exits,
     screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending DeviceBackupOperation.
     Resolves with string "stopped polling for pending Device Backups" if polling is stopped successfully; resolves with the empty string otherwise.
     */
    public func stopPollingForPendingDeviceBackups() -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.stopPollingPendingDeviceBackups(goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }

    /**
     Initiates an operation to add a Device to the DeviceGroup. Resolves with the operation name on successful initiation; rejects with
     an error otherwise.
     */
    public func addDevice(_ deviceGroup: String, device: String) -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.addDevice(
                    deviceGroup, device: device, receiver: goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }

    /**
     Polls for pending Devices (i.e. AddDeviceOperations), and returns the first set that materializes.
     Only one DeviceGroup can be polled at a time; thus, this function must return (by calling either
     stopPollingForPendingDevices or computeAddDeviceMPCOperation) before another call is made to this function.
     Resolves with a list of the pending Devices on success; rejects with an error otherwise.
     
     The "Operation" and "MPCData" keys are relevant on each "pending device".
        Operation: The mpc operation id for this pending device
        MPCData: the provided mpc data for adding this device (to use with MPCSdk.computeAddDeviceMPCOperation)
     */
    public func pollForPendingDevices(_ deviceGroup: String, pollInterval: NSNumber) -> Future<[PendingDevice], WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                do {
                    let pendingDeviceData = try self.keyClient?.pollPendingDevices(
                        deviceGroup,
                        pollInterval: pollInterval.int64Value)
                    let res = try JSONDecoder().decode([PendingDevice].self, from: pendingDeviceData!)
                    promise(Result.success(res))
                } catch {
                    promise(Result.failure(WaasError.mpcKeyServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }

    /**
     Stops polling for pending AddDevice operations. This function should be called, e.g., before your app exits,
     screen changes, etc. This function is a no-op if the SDK is not currently polling for a pending AddDeviceOperation.
     Resolves with string "stopped polling for pending Devices" if polling is stopped successfully; resolves with the empty string otherwise.
     */
    public func stopPollingForPendingDevices() -> Future<String, WaasError> {
        return Future { promise in
            Job.background().async(execute: {
                self.keyClient?.stopPollingPendingDevices(goReturnsString(promise: promise, wrapAsError: self.wrapError))
            })
        }
    }
}
